PROTOCOL.md

```diff
@@ -995,6 +1065,178 @@ Within the checkpoint, the `remove` struct does not contain the `stats` and `tag
+```
+
+For [safety reasons](#problems-with-multi-part-checkpoints), multi-part checkpoints MUST be clustered by
+spark-style hash partitioning. If the table supports [Deletion Vectors](#deletion-vectors), the partitioning
+key is the logical file identifier `(path, dvId)`; otherwise the key is just `path` (not `(path, NULL)`). This
+ensures deterministic content in each part file in case of multiple attempts to write the files -- even when
+older and newer Delta clients race.
+
+##### Problems with multi-part checkpoints
+
+Because they cannot be written atomically, multi-part checkpoints have several weaknesses:
+
+1. A writer cannot validate the content of the just-written checkpoint before readers could start using it.
+
+2. Two writers who race to produce the same checkpoint (same version, same number of parts) can overwrite each other, producing an arbitrary mix of checkpoint part files. If an overwrite changes the content of a file in any way, the resulting checkpoint may not produce an accurate snapshot.
+
+3. Not amenable to performance and scalability optimizations. For example, there is no way to store skipping stats for checkpoint parts, nor to reuse checkpoint part files across multiple checkpoints.
+
+4. Multi-part checkpoints also bloat the _delta_log dir and slow down LIST.
+
+[UUID-named](#uuid-named-checkpoint) checkpoint (which follow [V2 spec](#v2-spec)) solves all
+of these problems and should be preferred over multi-part checkpoints. For this reason, Multi-part
+checkpoints are forbidden when [V2 Checkpoints table feature](#v2-checkpoint-table-feature) is enabled.
+
+### Handling Backward compatibility while moving to UUID-named v2 Checkpoints
+
+A UUID-named v2 Checkpoint should only be created by clients if the [v2 checkpoint table feature](#v2-checkpoint-table-feature) is enabled.
+When UUID-named v2 checkpoints are enabled, Writers should occasionally create a v2 [Classic Checkpoint](#classic-checkpoint)
+for maintaining compatibility with older clients. These classic checkpoints have the same content as the UUID-named v2 checkpoint
+file name, but older clients (those which do not support [v2 checkpoint table feature](#v2-checkpoint-table-feature)) will recognize the classic
+file name, allowing them to extract [Protocol](#protocol-evolution) and fail gracefully with an invalid protocol
+version error on v2-checkpoint enabled tables. Writers should create classic checkpoints often enough to allow older
+clients to discover them and fail gracefully.
+
+### Allowed combinations for `checkpoint spec` <-> `checkpoint file naming`
+
+Checkpoint Spec | [UUID-named](#uuid-named-checkpoint) | [classic](#classic-checkpoint) | [multi-part](#multi-part-checkpoint)
+-|-|-|-
+[V1](#v1-spec) | Invalid | Valid | Valid
+[V2](#v2-spec) | Valid | Valid | Invalid
+
+### Metadata Cleanup
+
+The _deltaLog directory could increase in size as more and more commits and checkpoints are accumulated.
```

Comment:
```suggestion
The _deltaLog directory grows over time as more and more commits and checkpoints are accumulated.
```

Reply:
The classic checkpoint created should be complete in itself. It is a kind of v2 checkpoint which is just following classic naming convention. It is valid checkpoint which even any new Delta reader could read and initialize a snapshot and return correct results.